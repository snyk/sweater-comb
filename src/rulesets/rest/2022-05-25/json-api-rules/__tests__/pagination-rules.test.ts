import { OpenAPIV3 } from "@useoptic/openapi-utilities";
import { RuleRunner, TestHelpers } from "@useoptic/rulesets-base";
import { context } from "../../__tests__/fixtures";
import { paginationRules } from "../pagination-rules";

const baseJson = TestHelpers.createEmptySpec();

describe("pagination rules", () => {
  test("fails when pagination endpoint does not specify all pagination parameters", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users": {
          get: {
            parameters: [
              {
                name: "starting_after",
                in: "query",
              },
              {
                name: "ending_before",
                in: "query",
              },
            ],
            responses: {},
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([paginationRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });

  test("fails when non-paginated endpoint specifies any pagination parameter", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users": {
          "x-snyk-resource-singleton": true,
          get: {
            parameters: [
              {
                name: "starting_after",
                in: "query",
              },
            ],
            responses: {},
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([paginationRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });

  test("fails when paginated endpoints do not specify links", () => {
    const afterJson = {
      ...baseJson,
      paths: {
        "/api/users": {
          get: {
            parameters: [
              {
                name: "starting_after",
                in: "query",
              },
              {
                name: "ending_before",
                in: "query",
              },
              {
                name: "limit",
                in: "query",
              },
            ],
            responses: {
              "200": {
                description: "",
                content: {
                  "application/vnd.api+json": {
                    schema: {
                      properties: {
                        nolinks: {
                          type: "object",
                          properties: {},
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    } as OpenAPIV3.Document;

    const ruleRunner = new RuleRunner([paginationRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);
    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });
});
