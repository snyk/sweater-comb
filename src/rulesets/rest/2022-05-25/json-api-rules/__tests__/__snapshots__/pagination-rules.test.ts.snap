// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination rules fails when non-paginated endpoint specifies any pagination parameter 1`] = `
[
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": "expected operation to not support pagination parameter starting_after",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "unsupported pagination parameters",
    "passed": false,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "unsupported pagination parameters",
    "passed": true,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "unsupported pagination parameters",
    "passed": true,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
]
`;

exports[`pagination rules fails when paginated endpoints do not specify links 1`] = `
[
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "pagination parameters",
    "passed": true,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "pagination parameters",
    "passed": true,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "pagination parameters",
    "passed": true,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
  {
    "change": {
      "added": {
        "contentType": "application/vnd.api+json",
        "flatSchema": {},
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inResponse": {
            "body": {
              "contentType": "application/vnd.api+json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "responses",
          "200",
          "application/vnd.api+json",
        ],
        "jsonPath": "/paths/~1api~1users/get/responses/200/content/application~1vnd.api+json",
        "kind": "body",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": "Expected a partial match",
    "exempted": false,
    "expected": "{"schema":{"properties":{"links":{}}}}",
    "isMust": true,
    "isShould": false,
    "name": "pagination links",
    "passed": false,
    "received": "{"schema":{"properties":{"nolinks":{"type":"object","properties":{}}}}}",
    "type": "added",
    "where": "GET /api/users response 200 response body: application/vnd.api+json",
  },
]
`;

exports[`pagination rules fails when pagination endpoint does not specify all pagination parameters 1`] = `
[
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "pagination parameters",
    "passed": true,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "pagination parameters",
    "passed": true,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
  {
    "change": {
      "added": {
        "method": "get",
        "pathPattern": "/api/users",
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/jsonapi.md#pagination-parameters",
    "error": "Could not find a partial match in query parameters",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "pagination parameters",
    "passed": false,
    "received": undefined,
    "type": "added",
    "where": "GET /api/users",
  },
]
`;
